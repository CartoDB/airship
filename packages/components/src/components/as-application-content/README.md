# as-application-content

This component manages the layout of the application. It searches for our layout elements and manages for you the responsiveness behaviour, creating the correspondant panels. See `Layout` section for more information on our layout elements.

## How to use it

Add a `<as-application-content>` element as a wrapper for the application sections.

The component searches for these classes in order to know how many tabs it must render:
- `as-map-wrapper`.
- `as-sidebar`.
- `as-panels`.
- `as-bottom-bar`.

As long as you follow the proper markup, as the `Layout` section shows, you won't have to worry about anything else.

### Setting the tab names

In order to set the tabs text, you need to add a data attribute with the desired name:

`<aside class="as-sidebar as-sidebar--left" data-name="Stores">`

`<div class="as-panels" data-name="Legends">`

`<aside class="as-sidebar as-sidebar--right" data-name="Containers">`

You can change that data attribute programatically if you need to update the responsive tab name.

### Example

```html
noSource: true
responsive: true
---
<iframe src="/examples/components/as-application-content/simple.html" style="width: 100%; height: 100%;">
```

```code
lang: html
showSource: false
---
<as-application-content>
  <aside class="as-sidebar as-sidebar--left" data-name="Stores">
    LEFT-SIDEBAR
  </aside>

  <div class="as-map-wrapper">
    <div class="as-map">
      <div id="map"></div>
      <div class="as-panels" data-name="Legends">
        <div class="as-panel as-panel--top as-panel--right">
          <div class="as-panel__element">
            <p>Title</p>
            <p>
              Lorem ipsum dolor sit amet consectetur adipisicing elit. Inventore possimus nam quam necessitatibus omnis, est sequi rem
              quibusdam molestiae, at saepe odit voluptatem quae debitis consequatur iste ducimus quaerat dolorum.
            </p>
          </div>
          <div class="as-panel__element"> PANEL 1 </div>
        </div>

        <div class="as-panel as-panel--middle as-panel--right as-panel--vertical">
          <div class="as-panel__element">
            <p>Title</p>
            <p>
              Lorem ipsum dolor sit amet consectetur adipisicing elit. Inventore possimus nam quam necessitatibus omnis, est sequi rem
              quibusdam molestiae, at saepe odit voluptatem quae debitis consequatur iste ducimus quaerat dolorum.
            </p>
          </div>
          <div class="as-panel__element"> PANEL 1 </div>
        </div>
      </div>
    </div>

    <div class="as-bottom-bar">
      Lorem ipsum dolor sit amet consectetur adipisicing elit. Explicabo delectus ullam laudantium necessitatibus pariatur repellendus
      nemo ipsam minus. Delectus quidem sequi recusandae culpa deleniti molestias amet consequatur aspernatur eos exercitationem.
    </div>
  </div>

  <aside class="as-sidebar as-sidebar--right" data-name="Containers">
    RIGHT-SIDEBAR
  </aside>
</as-application-content>
```

### Changing the tab order

If you want to change the tabs order, you can add a data attribute to tell the component their positions.

Let's put `Legends` in the first position, `Stores` in the second and `Containers` in third place.

```code
lang: html
showSource: false
---
<div class="as-map-wrapper" data-tab-order="0">

<div class="as-panels" data-name="Legends" data-tab-order="1">

<aside class="as-sidebar as-sidebar--left" data-name="Stores" data-tab-order="2">

<aside class="as-sidebar as-sidebar--right" data-name="Containers" data-tab-order="3">
```

We recommend setting the map always at the position `0`.

```html
noSource: true
responsive: true
---
<iframe src="/examples/components/as-application-content/custom-order.html" style="width: 100%; height: 100%;">
```

## Events

### **load**

This event fires when the component has finished loading. Useful to load Leaflet with the proper content width and height.

```code
lang: javascript
---
const applicationContent = document.querySelector('as-application-content');
applicationContent.addEventListener('load', () => {});
```

```hint|directive
To avoid errors with layout dependent JavaScript code, we recommend you to execute all your JavaScript code inside the load event.
```

So, if you want to load a Leaflet map with some CARTO.js layers, you better put that code inside load event handler, like this:
```code
lang: javascript
---
const applicationContent = document.querySelector('as-application-content');
applicationContent.addEventListener('load', () => {
  const map = L.map('map', {
    zoomControl: false
  }).setView([30, 0], 3);
  map.scrollWheelZoom.disable();

  L.tileLayer('https://{s}.basemaps.cartocdn.com/rastertiles/voyager_nolabels/{z}/{x}/{y}.png', {
    maxZoom: 18
  }).addTo(map);

  const client = new carto.Client({
    apiKey: 'default_public',
    username: 'cartojs-test'
  });

  const source = new carto.source.Dataset('ne_10m_populated_places_simple');
  const style = new carto.style.CartoCSS(`
    #layer {
      marker-width: 7;
      marker-fill: #EE4D5A;
      marker-line-color: #FFFFFF;
    }
  `);
  const layer = new carto.layer.Layer(source, style);

  client.addLayer(layer);
  client.getLeafletLayer().addTo(map);
});
```

### **sectionChanged**

This event fires when the content section is changed via clicking a tab or via the programmatic method.

```code
lang: javascript
---
const applicationContent = document.querySelector('as-application-content');
applicationContent.addEventListener('sectionChanged', (section) => { console.log(section); });
```

## Methods

### **getSections()**

This method return a list of the sections discovered throughout the content with some interesting properties, such as:

```code
lang: typescript
---
interface ApplicationSection {
  active?: boolean;
  activeAction?: (section: ApplicationSection) => {};
  activeClass?: string;
  disableAction?: (section: ApplicationSection) => {};
  element: HTMLElement;
  name: string;
  type: string;
  tabOrder: number;
}
```

### **setVisible(sectionName)**

This method changes visible content section by introducing the section name like:

```code
lang: javascript
showSource: false
---
const applicationContent = document.querySelector('as-application-content');
aplicationContent.setVisible('Histogram');
```
