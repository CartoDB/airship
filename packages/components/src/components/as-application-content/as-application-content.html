<!DOCTYPE html>
<html dir="ltr" lang="en">
  <head>
    <base href="/">
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0, minimum-scale=1.0, maximum-scale=5.0">
    <title>Airship</title>

    <!-- CARTO.js -->
    <!-- Include Leaflet -->
    <script src="https://unpkg.com/leaflet@1.3.1/dist/leaflet.js"></script>
    <link href="https://unpkg.com/leaflet@1.3.1/dist/leaflet.css" rel="stylesheet">
    <link href="https://carto.com/developers/carto-js/examples/maps/public/style.css" rel="stylesheet">
    <!-- Include CARTO.js -->
    <script src="https://libs.cartocdn.com/carto.js/v4.1.2/carto.min.js"></script>

    <link rel="stylesheet" href="/dist/airship.css">
    <script src="/build/airship.js"></script>
  </head>

  <body class="as-app">
    <as-application-content>
      <aside class="as-sidebar as-sidebar--left">
        LEFT-SIDEBAR
      </aside>

      <div class="as-map-wrapper">
        <div class="as-map">
          <div id="map"></div>
          <div class="as-panels">
            <div class="as-panel as-panel--top as-panel--right">
              <div class="as-panel__element">
                <p>Title</p>
                <p>
                  Lorem ipsum dolor sit amet consectetur adipisicing elit. Inventore possimus nam quam necessitatibus omnis, est sequi rem
                  quibusdam molestiae, at saepe odit voluptatem quae debitis consequatur iste ducimus quaerat dolorum.
                </p>
              </div>
              <div class="as-panel__element"> PANEL 1 </div>
            </div>

            <div class="as-panel as-panel--middle as-panel--right as-panel--vertical">
              <div class="as-panel__element">
                <p>Title</p>
                <p>
                  Lorem ipsum dolor sit amet consectetur adipisicing elit. Inventore possimus nam quam necessitatibus omnis, est sequi rem
                  quibusdam molestiae, at saepe odit voluptatem quae debitis consequatur iste ducimus quaerat dolorum.
                </p>
              </div>
              <div class="as-panel__element"> PANEL 1 </div>
            </div>
          </div>
        </div>

        <div class="as-bottom-bar">
          Lorem ipsum dolor sit amet consectetur adipisicing elit. Explicabo delectus ullam laudantium necessitatibus pariatur repellendus
          nemo ipsam minus. Delectus quidem sequi recusandae culpa deleniti molestias amet consequatur aspernatur eos exercitationem.
        </div>
      </div>

      <aside class="as-sidebar as-sidebar--right">
        RIGHT-SIDEBAR
      </aside>
    </as-application-content>

    <script type="text/javascript">
      const applicationContent = document.querySelector('as-application-content');
      applicationContent.addEventListener('load', () => {
        const map = L.map('map', {
          zoomControl: false
        }).setView([30, 0], 3);
        map.scrollWheelZoom.disable();

        L.tileLayer('https://{s}.basemaps.cartocdn.com/rastertiles/voyager_nolabels/{z}/{x}/{y}.png', {
          maxZoom: 18
        }).addTo(map);

        const client = new carto.Client({
          apiKey: 'default_public',
          username: 'cartojs-test'
        });

        const source = new carto.source.Dataset('ne_10m_populated_places_simple');
        const style = new carto.style.CartoCSS(`
                #layer {
                  marker-width: 7;
                  marker-fill: #EE4D5A;
                  marker-line-color: #FFFFFF;
                }
              `);
        const layer = new carto.layer.Layer(source, style);

        client.addLayer(layer);
        client.getLeafletLayer().addTo(map);
      });
    </script>
  </body>
</html>
