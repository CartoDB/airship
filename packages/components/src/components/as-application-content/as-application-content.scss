@import '../../../../styles/src/core/_dev';

as-application-content {
  display: flex;
  flex: 1;
  flex-direction: column;
}
