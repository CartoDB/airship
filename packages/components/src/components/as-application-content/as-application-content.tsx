import { Component, Element, Event, EventEmitter, Method, State } from '@stencil/core';

/**
 * Category Widget
 *
 * @export
 * @class CategoryWidget
 */
@Component({
  shadow: false,
  styleUrl: './as-application-content.scss',
  tag: 'as-application-content'
})
export class ApplicationContent {
  @Element() private element: HTMLElement;

  @State() private sections: ApplicationSection[] = [];

  @Event() private load: EventEmitter<void>;
  @Event() private sectionChange: EventEmitter<object>;

  private activeSection: ApplicationSection;

  public render() {
    return [
      this.renderTabs(),
      this.renderContent()
    ];
  }

  public renderContent() {
    return (
      <main class='as-app-content'>
        <slot />
      </main>
    );
  }

  public componentWillLoad() {
    this.sections = this.getContentSections();
  }

  public componentDidLoad() {
    this.load.emit();
  }

  @Method()
  public getSections(): object[] {
    return this.sections;
  }

  @Method()
  public setVisible(sectionName: string) {
    const sectionFound = this.sections.find((section) => section.name === sectionName);

    if (sectionFound) {
      this.setActive(sectionFound);
    }
  }

  private renderTabs() {
    const tabs = this.sections.map((section, index) => {
      if (!section.element) {
        return;
      }

      const cssClasses = {
        'as-tabs__item': true,
        'as-tabs__item--active': section.active
      };

      return (
        <button onClick={() => this.setActive(section)} role='tab' class={cssClasses}>
          {section.name || index}
        </button>
      );
    });

    return (
      <div role='tablist' class='as-toolbar-tabs as-tabs as-tabs--xl'>{tabs}</div>
    );
  }

  private setActive(section: ApplicationSection) {
    if (section.active) {
      return;
    }

    this.disableActiveSection();

    if (section.activeAction) {
      section.activeAction(section);
    } else {
      section.element.classList.add(section.activeClass || ACTIVE_CLASSES[section.type]);
      section.active = true;
    }

    this.activeSection = section;
    this.sections = [...this.sections];
    this.sectionChange.emit(section);
  }

  private disableActiveSection() {
    if (!this.activeSection) {
      return;
    }

    if (this.activeSection.disableAction) {
      this.activeSection.disableAction(this.activeSection);
      return;
    }

    this.activeSection.element.classList.remove(
      this.activeSection.activeClass || ACTIVE_CLASSES[this.activeSection.type]
    );
    this.activeSection.active = false;
  }

  private getContentSections() {
    const sections = [
      this.getMap(),
      ...this.getSidebars(),
      ...this.getPanels(),
      this.getBottomBar()
    ];

    if (sections.length) {
      sections.sort((a, b) => a.tabOrder - b.tabOrder);
      this.setActive(sections[0]);
    }

    return sections;
  }

  private getMap() {
    const mapWrapper = this.element.querySelector('.as-map-wrapper');

    function activeAction(section: ApplicationSection) {
      section.active = true;
    }

    function disableAction(section: ApplicationSection) {
      section.active = false;
    }

    return {
      activeAction,
      disableAction,
      element: mapWrapper,
      name: mapWrapper.getAttribute('data-name') || 'Map',
      tabOrder: mapWrapper.getAttribute('data-tab-order') || 0,
      type: 'map'
    } as ApplicationSection;
  }

  private getSidebars() {
    const sidebars = Array.from(this.element.querySelectorAll('.as-sidebar'));

    const sidebarSections = sidebars.map((sidebar, index) => {
      return {
        activeClass: `as-sidebar--visible`,
        element: sidebar,
        name: sidebar.getAttribute('data-name') || `Sidebar ${index}`,
        tabOrder: sidebar.getAttribute('data-tab-order') || 0,
        type: 'sidebar'
      };
    }) as ApplicationSection[];

    return sidebarSections;
  }

  private getPanels() {
    const panels = Array.from(this.element.querySelectorAll('.as-panels'));

    const panelsSections = panels.map((panel, index) => ({
        element: panel,
        name: panel.getAttribute('data-name') || `Panel ${index}`,
        tabOrder: panel.getAttribute('data-tab-order') || 0,
        type: 'panels'
      })
    ) as ApplicationSection[];

    return panelsSections;
  }

  private getBottomBar() {
    const bottomBar = this.element.querySelector('.as-bottom-bar');

    return {
      element: bottomBar,
      name: bottomBar && bottomBar.getAttribute('data-name') || 'Bottom Bar',
      tabOrder: bottomBar && bottomBar.getAttribute('data-tab-order') || 0,
      type: 'bottomBar'
    } as ApplicationSection;
  }
}

const ACTIVE_CLASSES = {
  bottomBar: 'as-bottom-bar--visible',
  panels: 'as-panels--visible'
};

interface ApplicationSection {
  active?: boolean;
  activeAction?: (section: ApplicationSection) => {};
  activeClass?: string;
  disableAction?: (section: ApplicationSection) => {};
  element: HTMLElement;
  name: string;
  type: string;
  tabOrder: number;
}
