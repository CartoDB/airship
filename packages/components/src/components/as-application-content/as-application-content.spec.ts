import { TestWindow } from '@stencil/core/dist/testing';
import { ApplicationContent } from './as-application-content';

describe('as-category-widget', () => {
  it('should build', () => {
    expect(new ApplicationContent()).toBeTruthy();
  });

  describe('Rendering', async () => {
    let element: HTMLAsApplicationContentElement;
    let testWindow: TestWindow;

    beforeEach(async () => {
      testWindow = new TestWindow();
      element = await testWindow.load({
        components: [ApplicationContent],
        html: `<as-application-content>${DOMExample}</as-application-content>`
      });

      await testWindow.flush();
    });

    it('should render properly', () => {
      expect(element).toMatchSnapshot();
    });

    it('should render tabs based on content', () => {
      const tabs = element.querySelector('.as-toolbar-tabs');
      const tabsHTML = tabs.innerHTML;

      expect(tabsHTML).toContain('Sidebar 0');
      expect(tabsHTML).toContain('Sidebar 1');
      expect(tabsHTML).toContain('Panel 0');
      expect(tabsHTML).toContain('Bottom Bar');
    });

    it('should render child content inside .as-app-content', () => {
      const asAppContentNode = element.querySelector('.as-app-content');
      expect(asAppContentNode.innerHTML).toContain(DOMExample);
    });
  });

  describe('Behaviour', () => {
    let element: HTMLAsApplicationContentElement;
    let testWindow: TestWindow;

    beforeEach(async () => {
      testWindow = new TestWindow();
      element = await testWindow.load({
        components: [ApplicationContent],
        html: `<as-application-content>${DOMExample}</as-application-content>`
      });

      await testWindow.flush();
    });

    it('should change active content when clicking tabs', async () => {
      const tabs = element.querySelectorAll('.as-tabs__item');

      let mapTab = tabs[0];
      expect(mapTab).toMatchClasses(['as-tabs__item', 'as-tabs__item--active']);

      let leftSidebarTab = element.querySelectorAll('.as-tabs__item')[1] as HTMLElement;
      leftSidebarTab.click();
      await testWindow.flush();

      const tabsAfterClick = element.querySelectorAll('.as-tabs__item');
      const leftSidebarNode = element.querySelector('.as-sidebar.as-sidebar--left');

      mapTab = tabsAfterClick[0];
      leftSidebarTab = tabsAfterClick[1] as HTMLElement;

      expect(mapTab).toMatchClasses(['as-tabs__item']);
      expect(leftSidebarTab).toMatchClasses(['as-tabs__item', 'as-tabs__item--active']);
      expect(leftSidebarNode).toMatchClasses(['as-sidebar', 'as-sidebar--left', 'as-sidebar--left--visible']);
    });
  });

  describe('Events', async () => {
    let element: HTMLAsApplicationContentElement;
    let testWindow: TestWindow;

    beforeEach(async () => {
      testWindow = new TestWindow();
      element = await testWindow.load({
        components: [ApplicationContent],
        html: `<as-application-content>${DOMExample}</as-application-content>`
      });
    });

    it('should emit a "sectionChange" event when content changes in mobile view', async () => {
      const sectionChangeSpy = jest.fn();
      element.addEventListener('sectionChange', sectionChangeSpy);

      const leftSidebarTab = element.querySelectorAll('.as-tabs__item')[1] as HTMLElement;
      leftSidebarTab.click();

      await testWindow.flush();

      expect(sectionChangeSpy).toHaveBeenCalled();
    });
  });
});

const DOMExample = `
  <aside class="as-sidebar as-sidebar--left">Left Sidebar</aside>

  <div class="as-map-wrapper">
    <div class="as-map">
      <div id="map"></div>
      <div class="as-panels">
        <div class="as-panel as-panel--top as-panel--right">
          <div class="as-panel__element">Floating Panel</div>
        </div>
      </div>
    </div>

    <div class="as-bottom-bar">Bottom Bar</div>
  </div>

  <aside class="as-sidebar as-sidebar--right">Right Sidebar</aside>
`;
