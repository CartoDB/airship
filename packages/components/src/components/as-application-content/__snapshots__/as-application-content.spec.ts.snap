// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`as-category-widget Rendering should render properly 1`] = `
<as-application-content
  class="hydrated"
>
  
  
  
  
  
  
  
  
  <div
    class="as-toolbar-tabs as-tabs as-tabs--xl"
    role="tablist"
  >
    <button
      class="as-tabs__item as-tabs__item--active"
      role="tab"
    >
      Map
    </button>
    <button
      class="as-tabs__item"
      role="tab"
    >
      Sidebar 0
    </button>
    <button
      class="as-tabs__item"
      role="tab"
    >
      Sidebar 1
    </button>
    <button
      class="as-tabs__item"
      role="tab"
    >
      Panel 0
    </button>
    <button
      class="as-tabs__item"
      role="tab"
    >
      Bottom Bar
    </button>
  </div>
  <main
    class="as-app-content"
  >
    
    
  
    <aside
      class="as-sidebar as-sidebar--left"
    >
      Left Sidebar
    </aside>
    

  
    <div
      class="as-map-wrapper"
    >
      
    
      <div
        class="as-map"
      >
        
      
        <div
          id="map"
        />
        
      
        <div
          class="as-panels"
        >
          
        
          <div
            class="as-panel as-panel--top as-panel--right"
          >
            
          
            <div
              class="as-panel__element"
            >
              Floating Panel
            </div>
            
        
          </div>
          
      
        </div>
        
    
      </div>
      

    
      <div
        class="as-bottom-bar"
      >
        Bottom Bar
      </div>
      
  
    </div>
    

  
    <aside
      class="as-sidebar as-sidebar--right"
    >
      Right Sidebar
    </aside>
    

  </main>
</as-application-content>
`;
